files/routes/chat.py

```diff
@@ -1,117 +1,271 @@
+		def wrapper(*args: Any, **kwargs: Any) -> bool | None:
+			v = get_logged_in_user()
+			if not v:
+				return abort(403)
+			if not v.can_access_chat:
+				return abort(403)
+			if v.admin_level < perm_level:
+				return abort(403)
+			kwargs['v'] = v
+			return func(*args, **kwargs)
+		return wrapper
+	return wrapper_maker
+
+commands = {}
+
+def register_command(cmd_name, permission_level = 0):
+	def decorator(func):
+		@functools.wraps(func)
+		def wrapper(*args: Any, **kwargs: Any) -> bool | None:
+			v = get_logged_in_user()
+			if v.admin_level < permission_level:
+				send_system_reply(f"Unknown command: {cmd_name}")
+				return False
+			return func(*args, **kwargs)
+    
+		commands[cmd_name] = wrapper
+		
+		return wrapper
+	return decorator
+
+if app.debug:
+	socketio = SocketIO(
+		app,
+		async_mode='gevent',
+		logger=True,
+		engineio_logger=True,
+		debug=True,
+	)
 else:
-	socketio = SocketIO(app, async_mode='gevent', cors_allowed_origins=[SITE_FULL])
+	socketio = SocketIO(
+		app,
+		async_mode='gevent',
+	)
 
-typing = []
-online = []
-muted = cache.get(f'{SITE}_muted') or {}
-messages = cache.get(f'{SITE}_chat') or []
-total = cache.get(f'{SITE}_total') or 0
+CHAT_SCROLLBACK_ITEMS: Final[int] = 500
 
-@app.get("/chat")
-@auth_required
-def chat(v):
-	return render_template("chat.html", v=v, messages=messages)
+typing: list[str] = []
+online: list[str] = []	# right now we maintain this but don't actually use it anywhere
+connected_users = set()
+
+def send_system_reply(text):
+	data = {
+		"id": str(uuid.uuid4()),
+		"avatar": g.db.query(User).filter(User.id == NOTIFICATIONS_ID).one().profile_url,
+		"user_id": NOTIFICATIONS_ID,
+		"username": "System",
+		"text": text,
+		"text_html": sanitize(text),
+		'time': time.time(),
+	}
+	emit('speak', data)
+
+def get_chat_messages():
+	# Query for the last visible chat messages
+	result = (g.db.query(ChatMessage)
+          .join(User, User.id == ChatMessage.author_id)  # Join with the User table to fetch related user data
+          .order_by(ChatMessage.created_datetimez.desc())
+          .limit(CHAT_SCROLLBACK_ITEMS)
+          .all())
+	
+	# Convert the list of ChatMessage objects into a list of dictionaries
+	# Also, most recent at the bottom, not the top.
+	messages = [item.json_speak() for item in result[::-1]]
+	
+	return messages
 
+def get_chat_userlist():
+	# Query for the User.username column for users with chat_authorized == True
+    result = g.db.query(User.username).filter(User.chat_authorized == True).all()
+    
+    # Convert the list of tuples into a flat list of usernames
+    userlist = [item[0] for item in result]
+    
+    return userlist
 
-@app.get('/chat.js')
-@limiter.exempt
-def chatjs():
-	resp = make_response(send_from_directory('assets', 'js/chat.js'))
-	return resp
+@app.get("/chat")
+@is_not_permabanned
+@chat_is_allowed()
+def chat(v):
+	return render_template("chat.html", v=v)
 
 
 @socketio.on('speak')
 @limiter.limit("3/second;10/minute")
```

Comment:
10/minute sounds too harsh?

Reply:
Yup, good catch. (that one's not my fault, it was like that when I got here)